File: 2023/01/solution_part1.lua
-- local file = io.open("test_input_1.txt", "r")
local file = io.open("puzzle_input.txt", "r")
if not file then error("File not found") end

local sum = 0
for line in file:lines() do
    local digits = {}
    for digit in string.gmatch(line, "%d") do
        table.insert(digits, tonumber(digit, 10))
    end

    local number_of_digits = #digits
    if number_of_digits == 1 then
        sum = sum + digits[1] * 10 + digits[1]
    elseif number_of_digits == 2 then
        sum = sum + digits[1] * 10 + digits[2]
    elseif number_of_digits > 2 then
        sum = sum + digits[1] * 10 + digits[number_of_digits]
    end
end
print(sum)

File: 2023/01/solution_part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func main() {
	// f, err := os.Open("test_input_1.txt")
	f, err := os.Open("puzzle_input.txt")
	if err != nil {
		fmt.Println(err)
		return
	}
	defer f.Close()

	sum := 0
	s := bufio.NewScanner(f)
	for s.Scan() {
		sum += process_line(s.Text())
	}
	fmt.Println(sum)
}

func process_line(line string) int {
	// remove all non-digits from string
	re := regexp.MustCompile(`\D`)
	digits_from_line := re.ReplaceAllString(line, "")

	length := len(digits_from_line)
	if length == 0 {
		return 0
	}

	tens := digits_from_line[0]
	ones := digits_from_line[length-1]
	result := string([]byte{tens, ones})
	parsed, err := strconv.Atoi(result)
	if err != nil {
		panic(err)
	}
	return parsed
}
